src/operator/correlation-inl.h

```diff
@@ -228,6 +228,22 @@ void Init(const std::vector<std::pair<std::string, std::string> >& kwargs) overr
     out_shape->push_back(Shape4(dshape1[0], paddedbottomheight, paddedbottomwidth, dshape1[1]));
     return true;
   }
+  bool InferType(std::vector<int> *in_type,
+                 std::vector<int> *out_type,
+                 std::vector<int> *aux_type) const override {
+    int dtype = (*in_type)[0];
+    type_assign(&(*in_type)[1], dtype);
+    type_assign(&(*out_type)[0], dtype);
+    type_assign(&(*out_type)[1], dtype);
+    type_assign(&(*out_type)[2], dtype);
+
+    TYPE_ASSIGN_CHECK(*in_type, 0, dtype);
```

Comment:
Why do you need both type_assign and TYPE_ASSIGN_CHECK ?

Reply:
It was advice from Jun, he suggested that we can do a mutual inference by reducing all datatypes to one and then assign the reduced type back to everything.

Reply:
If the reduced datatype is not -1 then all inputs and outputs will then have the same datatype at the end of this function.